### DISCUSSION TITLE
PorousFlow: EOS for water, air, with vapor pressure lowering capability

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#23280] [Q&A Modules: General] [answered]
Hello,

I’m looking to simulate a multiphase (liquid and gas) and multicomponent flow with water and air as the components. I’d like to know which fluid EOS in PorousFlow is appropriate for this simulation. My search results reveal several fluid EOS, including that of water, but I don’t see any EOS relating to the fluid properties of air. This simulation is, however, possible using the EOS4 module of TOUGH2. Is there an alternative for the TOUGH2 EOS4 module in PorousFlow? Thanks.

**Comment by cpgr** [comment, marked as answer]
I don't believe that there is an air EOS yet, and I don't think that we have vapor pressure lowering like in TOUGH2's EOS4. I think from memory that EOS4 just uses an ideal gas for air.

**Reply by batodon** [reply]
Thanks a lot, @cpgr. It’s been a minute. Very refreshing to get feedback from you again! In my intended simulation, the liquid phase contains `water` and `dissolved air` as components, while the gaseous phase comprises `air` and `water vapor` as components. Please:
1.	Do you believe I can achieve this simulation in PorousFlow using the fluid properties of water and ideal gas? 
2.	Will you advise that I implement/add the TOUGH2 EOS4 Module in PorousFlow? Or will this implementation take a while?

Thanks again!

**Reply by cpgr** [reply]
I think that you could do 1 with only a very small amount of coding (basically making an AirFluidProperties that is an ideal gas but with Henry's law coefficients supplied, which IdealGasFluidProperties doesn't have) if an ideal gas representation is ok for your conditions. If not, it would be a bit more effort to implement an air EOS - not difficult, but tedious.

To add vapour pressure lowering will require another bit of code, but it probably wouldn't take too much effort either.

Both would be useful additions to MOOSE. If you are happy to have a go at this I'm happy to help.

**Reply by batodon** [reply]
Thanks, @cpgr. 
I’ll go with option 1. I’m keeping this ticket open while I work on it. I’ll get back to you shortly.

**Reply by batodon** [reply]
Hello @cpgr,
I supplied the Henry coefficient to the Ideal gas fluid property (I call this new fluid property the `IdealgasHenry`). I also created a `PorousflowWaterAir` fluid-state object based on the `PorousflowWaterNCG` object using this fluid property. The model ran well after I tested it on the `waterncg.i`. I’m, however, not entirely convinced about my implementation. Could you please offer a second pair of eyes on it for me? The code is here:

https://github.com/iyevugah/ERGS/tree/main/src/userobjects 

The only minor addition to the `PorousflowWaterAir` is from lines 130 to 136, where I included the vapour pressure-lowering capability as follows (not sure if it is entirely correct):

 ```
 // Vapor pressure (with lowering capability)
  DualReal numer = _Mh2o *_pc.capillaryPressure(liquid.saturation, qp) * liquid.saturation;
  DualReal deno  = liquid.density * _R * (T + 273.15);
  DualReal f_vpl = std::exp(numer/deno);
  DualReal pv = f_vpl * psat;

  // Save pressures to FluidStateProperties object
  gas.pressure = p + pv;

```
Please, see the problem folder for the input file. Also, the formula for the vapour pressure lowering effect is available on page 145 in the TOUGH2 Manual as equation A.7. You can get it from the link below: 
https://drive.google.com/file/d/1nFowvaOJTImLmsScGu3mB5p-6khVexsn/view?usp=share_link 

Thanks.

**Reply by cpgr** [reply]
It looks ok at first glance. One thing to check is that I think `T` is in Kelvin once it gets to `PorousflowWaterAir`, so you probably don't want to add 273.15 to it.

You could also dramatically reduce the amount of code by inheriting from `IdealGasFluidProperties` and `PorousFlowWaterNCG` and only overriding the bits you added. For example, `IdealGasHenryFluidProperties.h` could be something like

```
#pragma once

#include "IdealGasFluidProperties.h"

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Woverloaded-virtual"

/**
 * Ideal gas fluid properties
 * Default parameters are for air at atmospheric pressure and temperature
 */
class IdealGasHenryFluidProperties : public IdealGasFluidProperties, public NaNInterface
{
public:
  static InputParameters validParams();

  IdealGasHenryFluidProperties(const InputParameters & parameters);

  virtual std::vector<Real> henryCoefficients() const override;
};

#pragma GCC diagnostic pop
```

and similarly `IdealGasHenryFluidProperties.C`
```
#include "IdealGasHenryFluidProperties.h"

registerMooseObject("FluidPropertiesApp", IdealGasHenryFluidProperties);

InputParameters
IdealGasHenryFluidProperties::validParams()
{
  InputParameters params = IdealGasFluidProperties::validParams();

  return params;
}

IdealGasHenryFluidProperties::IdealGasHenryFluidProperties(const InputParameters & parameters)
  : IdealGasFluidProperties(parameters),
    NaNInterface(this) {}

std::vector<Real>
IdealGasHenryFluidProperties::henryCoefficients() const
{
  return {-8.55445, 4.01195, 9.52345};
}
```

(note: I actually haven't tried to compile this but it should be similar).

**Reply by batodon** [reply]
Great! Thanks a million, @cpgr!  I’ll make the code more succinct! 👍

**Reply by cpgr** [reply]
Let me know if you run into any issues but it shouldn't be too hard. The TOUGH2 manual shows some results using EOS3/EOS4 for one of the sample problems. It would be good to see if you could reproduce that at all.

**Reply by batodon** [reply]
Hi @cpgr, I have only one hurdle. I keep getting the following error which I’m trying to fix:

```
warning: direct base 'NaNInterface' inaccessible in ' IdealGasHenryFluidProperties
' due to ambiguity [-Winaccessible-base]
   12 | class IdealGasHenryFluidProperties : public IdealGasFluidProperties, public NaNInterface 
```

It seems I have a duplicate of `NaNInterface`, but when I removed it, I still got an error. Could you please help with this error?  

It will be great to validate the code with those tests. I’ll do that once I fix this error. Thanks.

**Reply by cpgr** [reply]
Oops, I was copying and pasting and should have deleted the NanInterface things I think. How about

```
#pragma once

#include "IdealGasFluidProperties.h"

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Woverloaded-virtual"

/**
 * Ideal gas fluid properties
 * Default parameters are for air at atmospheric pressure and temperature
 */
class IdealGasHenryFluidProperties : public IdealGasFluidProperties
{
public:
  static InputParameters validParams();

  IdealGasHenryFluidProperties(const InputParameters & parameters);

  virtual std::vector<Real> henryCoefficients() const override;
};

#pragma GCC diagnostic pop
```

and

```
#include "IdealGasHenryFluidProperties.h"

registerMooseObject("FluidPropertiesApp", IdealGasHenryFluidProperties);

InputParameters
IdealGasHenryFluidProperties::validParams()
{
  InputParameters params = IdealGasFluidProperties::validParams();

  return params;
}

IdealGasHenryFluidProperties::IdealGasHenryFluidProperties(const InputParameters & parameters)
  : IdealGasFluidProperties(parameters)
{
}

std::vector<Real>
IdealGasHenryFluidProperties::henryCoefficients() const
{
  return {-8.55445, 4.01195, 9.52345};
}
```

**Reply by cpgr** [reply]
Also, maybe `IdealGasAirFluidProperties` would be a better name?

**Reply by batodon** [reply]
Yeah, this works beautifully!

**Reply by batodon** [reply]
Thanks, @cpgr. I'll try the examples in the TOUGH2 manual and get back shortly. :)

**Reply by batodon** [reply]
Hello @cpgr, so, I finally got my simulation running, but the results were not quite what I expected (e.g., the temperature and pgas are constant throughout). I’m pasting my input file here. Also, I’d like to please iterate that I’m using the parameters on page 58 of the user manual provided by Petrasim®. I’m attaching the link for anyone who might want to reference it later. Could you please cross-check the input file for me? I believe I might be missing a few things. Also, in PorousFlow 2-phase 2-component simulations, the total diffusion coefficients should be 4, but I wonder what the best order to provide these coefficients is. For example, going by the diffusion coefficients on page 58, I entered `2.13e-5 0 2.13e-5 0`, but I don’t fully understand the pattern. Probably, it should be `0 2.13e-5 0 2.13e-5` Could you please clarify this for me? On line 114 in the input file, the function should read: log(x/sqrt(t)), but the input file is not accepting the logarithmic function. Thanks.

https://drive.google.com/file/d/18FI2iDvZylHzcjD-7pYokmwSoSeeXm2m/view?usp=share_link

#TOUGH2 EOS3/4 example

```
[Mesh]
  [efmcube]
  type = FileMeshGenerator
  file = eos4.msh
  []
[]


[FluidProperties]
  [air]
    type = IdealGasAirFluidProperties
  []
  [water]
    type = Water97FluidProperties
  []
[]


[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pgas z temperature'
    number_fluid_phases = 2
    number_fluid_components = 2
  []
  [pc]
    type = PorousFlowCapillaryPressureVG
    m = 0.45
    alpha = 1e1
    pc_max = 1e4
  []
  [fs]
    type = PorousFlowWaterAir
    water_fp = water
    gas_fp = air
    capillary_pressure = pc
  []
[]


[GlobalParams]
  PorousFlowDictator = dictator
  gravity = '0 0 0'
[]


[Variables]
  [pgas]
    initial_condition = 1.0E5
  []
  [z]
    initial_condition = 0.2
  []
  [temperature]
    initial_condition = 291.15
  []
[]


[AuxVariables]
  [sgas]
    family = MONOMIAL
    order = CONSTANT
  []
  [sLiquid]
    family = MONOMIAL
    order = CONSTANT
  []
  [xAir]
   order = CONSTANT
    family = MONOMIAL
  []
  [Z]
    order = FIRST
    family = LAGRANGE
  []
[]


[AuxKernels]
  [sgas]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 1
    variable = sgas
  []
  [sLiquid]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 0
    variable = sLiquid
  []
  [xAir]
    type = PorousFlowPropertyAux
    variable = xAir
    property = mass_fraction
    phase = 0
    fluid_component = 1
    execute_on = timestep_end
  []
 [Z]
    type = FunctionAux
    variable = Z
    function = sol_variable
 []
[]


[Functions]
  [sol_variable]
   type = ParsedFunction 
   value = (x/sqrt(t))
  []
[]

[Kernels]
  [mass0]
    type = PorousFlowMassTimeDerivative
   variable = pgas
    fluid_component = 0
  []
  [adv0]
    type = PorousFlowAdvectiveFlux
    variable = pgas
    fluid_component = 0
  []
   [disp0]
    type = PorousFlowDispersiveFlux
    disp_trans = '0 0'
    disp_long = '0 0'
    fluid_component = 0
    use_displaced_mesh = false
    variable = pgas
  []
  [mass1]
    type = PorousFlowMassTimeDerivative
    variable = z
    fluid_component = 1
  []
  [adv1]
    type = PorousFlowAdvectiveFlux
    variable = z
    fluid_component = 1
  []
   [disp1]
    type = PorousFlowDispersiveFlux
    disp_trans = '0 0'
    disp_long = '0 0'
    fluid_component = 1
    use_displaced_mesh = false
    variable = z
  []
  [energy]
    type = PorousFlowEnergyTimeDerivative
    variable = temperature
  []
  [heat_adv]
    type = PorousFlowHeatAdvection
    variable = temperature
  []  
  [conduction]
    type = PorousFlowHeatConduction
    variable = temperature
  []  
  [heat_source]
    type = HeatSource
    variable = temperature
    value = 3e3
  []
[]


[Materials]
  [temperature]
    type = PorousFlowTemperature
    temperature = temperature
  []
  [WaterAirProperties]
    type = PorousFlowFluidState
    gas_porepressure = pgas
    z = z
    temperature = temperature
    capillary_pressure = pc
    fluid_state = fs
  []
  [permeability]
    type = PorousFlowPermeabilityConst
    permeability = '20.e-15 0 0   0 20.e-15 0   0 0 20.e-15'
  []
  [relperm0]
    type = PorousFlowRelativePermeabilityVG
    m = 0.45
    phase = 0
  []
  [relperm1]
    type = PorousFlowRelativePermeabilityVG
    m = 0.45
    phase = 1
  []
  [porosity]
    type = PorousFlowPorosityConst
    porosity = 0.01
  []
  [rock_heat]
    type = PorousFlowMatrixInternalEnergy
    specific_heat_capacity = 800
    density = 2550
  [] 
  [diffusivity]
    type = PorousFlowDiffusivityConst
    diffusion_coeff = '2.13E-5 0 2.13E-5 0'
    tortuosity = '0.25 0.25'
  []
  [rock_thermal_conductivity]
    type = PorousFlowThermalConductivityIdeal
    dry_thermal_conductivity = '2 0 0  0 2 0  0 0 2'
  []
[]


[VectorPostprocessors]
  [variables]
    type = LineValueSampler
    variable = 'Z xAir pgas temperature sLiquid'
    sort_by = id
    start_point = '0 0 0'
    end_point = '100 0 0'
    num_points = 100
    execute_on = 'timestep_end'
  []
[]


[Postprocessors]
  [Z]
    type = PointValue
    point = '1 0 0'
    variable = temperature
  []
  [xAir]
    type = PointValue
    point =  '1 0 0'
    variable = xAir
  []
  [Pgas]
    type = PointValue
   point =  '1 0 0'
    variable = pgas
  []
  [T]
    type = PointValue
    point = '1 0 0'
    variable = temperature
  []
  [SLiquid]
    type = PointValue
    point =  '1 0 0'
    variable = sLiquid
  []
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]


[Executioner]
  type = Transient
  solve_type = NEWTON
  dt = 1
  end_time = 10
  nl_abs_tol = 1e-12
[]

[Outputs]
  exodus = true
  [csv]
  type = CSV
  execute_on = 'initial timestep_end'
  []
[]
```

**Reply by batodon** [reply]
Hello @cpgr, any updates on this issue, please? thanks.

**Reply by cpgr** [reply]
I made a start on this but am currently interstate for work all week so can't really do much until next week.

**Reply by batodon** [reply]
Great. Thanks for the feedback, @cpgr. While I wait to get an update, hopefully, next week, I’ll continue working on it. I’ll let you know if I get some interesting results, too. Thanks again.

**Reply by cpgr** [reply]
From memory I think that you want a bc for heat on the left edge, rather than a body source of heat everywhere.

**Reply by batodon** [reply]
oops. Sorry, I updated this input file with a `DiracKernel` that indicates a heat source in the first cell; I forgot to update it here. I’ll consider changing it to a bc, then. Thanks.  

```
[DiracKernels]
  [heat_source]
    type = PorousFlowSquarePulsePointSource
    point = '1 0 0'
    mass_flux = 3e3
    variable = temperature
  []
[]
```

**Reply by batodon** [reply]
Hello @cpgr, perhaps not the best time to update you on my progress, but I’m still going to put it out there so that you can consider it when you get back to work and want to update me. I attached below a short movie and the input file for my simulation. I also attached some graphs showing my results here. These graphs show the primary variables (i.e., pressure and temperature) versus the distance from the temperature source. The results differ based on how I set the diffusion coefficients. I called the different results scenario 1 (i.e., diffusion coefficients  = `2.13E-5  0  2.13E-5  0`)  and scenario 2 (diffusion coefficients  = `0  2.13E-5  0  2.13E-5`).

In both scenarios, the temperature graphs look more like the results from the benchmark example, except for the small squiggle line at the beginning of the ‘curve decline’ in the benchmark result. 
The pressure graph from scenario 1 looks entirely different from scenario 2, which appears to follow the benchmark results.    
Please, note that I’m still unsure if these results accurately represent the results from the benchmark. Thanks. 

simulation for scenario2
https://user-images.githubusercontent.com/113298585/218248824-b86d1aef-009c-4e1a-bea8-2f3a69c3acab.mp4

P&T Results for scenario 1
![s1p](https://user-images.githubusercontent.com/113298585/218249577-1b39f4e8-5311-42ea-8f12-db47d4c8aa2d.PNG)

![s1t](https://user-images.githubusercontent.com/113298585/218248868-9a1f4028-c32b-428e-a5cf-94aec95465a6.PNG)

P&T Results for scenario1

![s2p](https://user-images.githubusercontent.com/113298585/218249061-da75da76-2c29-4d55-9e1a-b2b057e2c1e7.PNG)
![s2t](https://user-images.githubusercontent.com/113298585/218249069-94173fa1-ee9e-4018-b3e3-06879169bfba.PNG)

**Reply by batodon** [reply]
input file:
```
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = 500   #500
  xmax = 10000 #1000
  ny = 1
  ymax = 4.5
  bias_x = 1.01
[]


[FluidProperties]
  [air]
    type = IdealGasAirFluidProperties
  []
  [water]
    type = Water97FluidProperties
  []
[]


[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pgas zi temperature'
    number_fluid_phases = 2
    number_fluid_components = 2
  []
  [pc]
    type = PorousFlowCapillaryPressureVG
    m = 0.45
    alpha = 1e1 
  []
  [fs]
    type = PorousFlowWaterAir
    water_fp = water
    gas_fp = air
    capillary_pressure = pc
  []
[]


[GlobalParams]
  PorousFlowDictator = dictator
  gravity = '0 0 0'
[]


[Variables]
  [pgas]
    initial_condition = 1.0E5
  []
  [zi]
    initial_condition = 0.01  #0.14 0.01 to make it converge
  []
  [temperature]
    initial_condition = 291.15
  []
[]


[AuxVariables]
  [sgas]
   order = CONSTANT
   family = MONOMIAL
  []
  [swater]
   order = CONSTANT
   family = MONOMIAL
  []
  [xAir]
   order = CONSTANT
    family = MONOMIAL
  []
[]


[AuxKernels]
  [sgas]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 1
    variable = sgas
  []
  [swater]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 0
    variable = swater
  []
  [xAir]
    type = PorousFlowPropertyAux
    variable = xAir
    property = mass_fraction
    phase = 0
    fluid_component = 1
    execute_on = timestep_end
  []
[]


[Kernels]
  [mass0]
    type = PorousFlowMassTimeDerivative
    variable = pgas
    fluid_component = 0
  []
  [adv0]
    type = PorousFlowAdvectiveFlux
    variable = pgas
    fluid_component = 0
  []
   [disp0]
    type = PorousFlowDispersiveFlux
    variable = pgas
    disp_trans = '0 0'
    disp_long = '0 0'
    fluid_component = 0
    use_displaced_mesh = false
  []
  [mass1]
    type = PorousFlowMassTimeDerivative
    variable = zi
    fluid_component = 1
  []
  [adv1]
    type = PorousFlowAdvectiveFlux
    variable = zi
    fluid_component = 1
  []
   [disp1]
    type = PorousFlowDispersiveFlux
    variable = zi
    disp_trans = '0 0'
    disp_long = '0 0'
    fluid_component = 1
    use_displaced_mesh = false
  []
  [energy]
    type = PorousFlowEnergyTimeDerivative
    variable = temperature
  []
  [heat_adv]
    type = PorousFlowHeatAdvection
    variable = temperature
  []  
  [conduction]
    type = PorousFlowHeatConduction
    variable = temperature
  []
[]


[Materials]
  [temperature]
    type = PorousFlowTemperature
    temperature = temperature
  []
  [WaterAirProperties]
    type = PorousFlowFluidState
    gas_porepressure = pgas
    z = zi
    temperature = temperature
    capillary_pressure = pc
    fluid_state = fs
  []
  [permeability]
    type = PorousFlowPermeabilityConst
    permeability = '20.e-15 0 0   0 20.e-15 0   0 0 20.e-15'
  []
  [relperm0]
    type = PorousFlowRelativePermeabilityVG
    m = 0.45
    phase = 0
  []
  [relperm1]
    type = PorousFlowRelativePermeabilityVG
    m = 0.45
    phase = 1
  []
  [porosity]
    type = PorousFlowPorosityConst
    porosity = 0.01
  []
  [rock_heat]
    type = PorousFlowMatrixInternalEnergy
    specific_heat_capacity = 800
    density = 2550
  [] 
  [diffusivity]
    type = PorousFlowDiffusivityConst
    diffusion_coeff = ' 0 2.13E-5 0  2.13E-5'
    tortuosity = '0.25 0.25'
  []
  [rock_thermal_conductivity]
    type = PorousFlowThermalConductivityIdeal
    dry_thermal_conductivity = '2 0 0  0 2 0  0 0 2'
  []
[]


[BCs]
  [flux]
    type = PorousFlowSink
    boundary = 'left'
    variable = temperature
    use_mobility = false
    use_relperm = false
    mass_fraction_component = 1
    fluid_phase = 0
    flux_function = -3e3
  []
[]


[VectorPostprocessors]
  [vars]
    type = NodalValueSampler
    sort_by = id
    variable = 'pgas temperature'
    execute_on = 'timestep_end'
    outputs = spatial
  []
  [auxvars]
    type = ElementValueSampler
    sort_by = id
    variable = 'xAir  swater'
    execute_on = 'timestep_end'
    outputs = spatial
  []
[]


[Postprocessors]
  [xAir]
    type = PointValue
    point =  '1 0 0'
    variable = xAir
  []
  [Pgas]
    type = PointValue
   point =  '1 0 0'
    variable = pgas
  []
  [T]
    type = PointValue
    point = '1 0 0'
    variable = temperature
  []
  [swater]
    type = PointValue
    point =  '1 0 0'
    variable = swater
  []
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]


[Executioner]
  type = Transient
  solve_type = NEWTON
  dt = 1e4
  num_steps = 1e6
  end_time = 3.56E8
  nl_abs_tol = 1e-20
  nl_max_its = 25
  l_max_its = 100
  dtmax = 5e6
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 10
  []
[]

[Outputs]
  exodus = true
  sync_times = '1e5 2.592e6 3.64e7 3e8'
  [time]
    type = CSV
  []
  [spatial]
    type = CSV
    sync_only = true
  []
[]
```

**Reply by cpgr** [reply]
Sorry that this has taken so long to reply - I didn't actually see this until a couple of days ago.

It's looking promising. A couple of things:

- This will have a similarity solution if you make it radial - you can add `coord_type = RZ` in the `[Mesh]` block
- To get the gas saturation = 0.2 initially, you want to use a `PorousFlowFluidStateIC` to compute the correct initial value for z
```
[ICs]
  [z]
    type = PorousFlowFluidStateIC
    variable = z
    saturation = 0.2
    gas_porepressure = pgas
    temperature = temperature
    fluid_state = fs
  []
[]
```
- Your `xAir` is the mass fraction of the air component in the liquid phase, but I think you want the mass fraction of air in the gas phase, so change `phase=1` in the `xAir AuxKernel`.
- You could do the non-vapour pressure lowering case (EOS3 in TOUGH2) by swapping your `PorousFlowWaterAir` to `PorousFlowWaterNCG`!
- The `diffusion_coeff` in the `PorousFlowDiffusivityConst` material is the following order: water in liquid phase, air in liquid phase, water in gas phase, air in gas phase
- I think the dirac kernel is probably the correct way to add heat energy in your case
- You could plot the values for fixed r but varying t (the post processors) or varying r but fixed t (the vector post processor) and they should coincide for the radial case like at https://mooseframework.inl.gov/modules/porous_flow/1Dradial.html

**Reply by batodon** [reply]
Hello @cpgr, this is awesome! I’ll make the necessary corrections and get back to you shortly! I appreciate this so much!

**Reply by batodon** [reply]
Hello @cpgr, no matter what I do, I keep getting the following error when I included the `ICs`:

```
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```

I guess it might be a mesh problem, although I highly doubt it.  I include my input mesh here. Any suggestions, pls? Thanks.

```
[Mesh]
 [gen]
  type = GeneratedMeshGenerator
  dim = 2
  nx = 500   #500
  xmax = 10000 #1000
  xmin = 1
  ny = 1
  ymax = 4.5
  ymin = 1
  bias_x = 1.01
  []
  coord_type = RZ
[]
```

**Reply by batodon** [reply]
Sorry @cpgr, Please, ignore the previous message. I found the bug! It turns out that I need to specify the `temperature_units` as a global param instead of specifying it under materials.

**Reply by batodon** [reply]
Hi @cpgr, I guess I’m probably missing something because it’s taking forever for my simulation to converge with intermittent notifications as follows:

```
A MooseException was raised during Auxiliary variable computation.
The next solve will fail, the timestep will be reduced, and we will try again.

fs: temperature 273.154 is outside range 273.16 K <= T <= 647.096 K
```
Any suggestions for a faster convergence, pls? I’m using the RZ coordinates, as you suggested. I also tried extrapolating the material properties using TabulatedBicubicFluidProperties. Thank you.

**Reply by cpgr** [reply]
I'm happy to take a look but probably need the current input file otherwise I'd be guessing!

**Reply by batodon** [reply]
ok, great @cpgr. Here it is:

```
[Mesh]
 [gen]
  type = GeneratedMeshGenerator
  dim = 2
  nx = 500   #500
  xmax = 10000 #1000
  xmin = 1
  ny = 1
  ymax = 4.5
  ymin = 1
  bias_x = 1.01
  []
  coord_type = RZ
[]


[FluidProperties]
  [true_water]
    type = Water97FluidProperties
  []
  [tabulated_water]
    type = TabulatedBicubicFluidProperties
    fp = true_water
    temperature_min = 250
    pressure_max = 1E8
    interpolated_properties = 'density viscosity enthalpy internal_energy'
    fluid_property_file = water97_tabulated_11.csv
  []
  [true_co2]
    type = CO2FluidProperties
  []
  [tabulated_co2]
    type = TabulatedBicubicFluidProperties
    fp = true_co2
    temperature_min = 250
    pressure_max = 1E8
    interpolated_properties = 'density viscosity enthalpy internal_energy'
    fluid_property_file = co2_tabulated_11.csv
  []
  [air]
    type = IdealGasAirFluidProperties
  []
[]

[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pgas z temperature'
    number_fluid_phases = 2
    number_fluid_components = 2
  []
  [pc]
    type = PorousFlowCapillaryPressureVG
    m = 0.45
    alpha = 1e1 
  []
  [fs]
    type = PorousFlowWaterNCG #PorousFlowWaterAir   #
    water_fp = true_water
    gas_fp = true_co2 #air                # 
    capillary_pressure = pc
  []
[]


[GlobalParams]
  PorousFlowDictator = dictator
  gravity = '0 0 0'
  temperature_unit = Celsius
[]


[Variables]
  [pgas]
    initial_condition = 1e5
  []
  [z]
  []
  [temperature]
    initial_condition = 18 # 291.15
  []
[]


[ICs]
  [z]
    type = PorousFlowFluidStateIC
    variable = z
    saturation = 0.2
    gas_porepressure = pgas
    temperature = temperature
    fluid_state = fs
  []
[]


[AuxVariables]
  [sgas]
   order = CONSTANT
   family = MONOMIAL
  []
  [swater]
   order = CONSTANT
   family = MONOMIAL
  []
  [xAir]
   order = CONSTANT
    family = MONOMIAL
  []
[]


[AuxKernels]
  [sgas]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 1
    variable = sgas
  []
  [swater]
    type = PorousFlowPropertyAux
    property = saturation
    phase = 0
    variable = swater
  []
  [xAir]
    type = PorousFlowPropertyAux
    variable = xAir
    property = mass_fraction
    phase = 1
    fluid_component = 1
    execute_on = timestep_end
  []
[]


[Kernels]
  [mass0]
    type = PorousFlowMassTimeDerivative
    variable = pgas
    fluid_component = 0
  []
  [adv0]
    type = PorousFlowAdvectiveFlux
    variable = pgas
    fluid_component = 0
  []
   [disp0]
    type = PorousFlowDispersiveFlux
    variable = pgas
    disp_trans = '0 0'
    disp_long = '0 0'
    fluid_component = 0
    use_displaced_mesh = false
  []
  [mass1]
    type = PorousFlowMassTimeDerivative
    variable = z
    fluid_component = 1
  []
  [adv1]
    type = PorousFlowAdvectiveFlux
    variable = z
    fluid_component = 1
  []
   [disp1]
    type = PorousFlowDispersiveFlux
    variable = z
    disp_trans = '0 0'
    disp_long = '0 0'
    fluid_component = 1
    use_displaced_mesh = false
  []
  [energy]
    type = PorousFlowEnergyTimeDerivative
    variable = temperature
  []
  [heat_adv]
    type = PorousFlowHeatAdvection
    variable = temperature
  []  
  [conduction]
    type = PorousFlowHeatConduction
    variable = temperature
  []
[]


[Materials]
  [temperature]
    type = PorousFlowTemperature
    temperature = temperature
  []
  [WaterAirProperties]
    type = PorousFlowFluidState
    gas_porepressure = pgas
    z = z
    temperature = temperature
    capillary_pressure = pc
    fluid_state = fs
  []
  [permeability]
    type = PorousFlowPermeabilityConst
    permeability = '20.e-15 0 0   0 20.e-15 0   0 0 20.e-15'
  []
  [relperm0]
    type = PorousFlowRelativePermeabilityVG
    m = 0.45
    phase = 0
  []
  [relperm1]
    type = PorousFlowRelativePermeabilityVG
    m = 0.45
    phase = 1
  []
  [porosity]
    type = PorousFlowPorosityConst
    porosity = 0.01
  []
  [diffusivity]
    type = PorousFlowDiffusivityConst
    diffusion_coeff = ' 2.13E-5 2.13E-5  0  0 '
    tortuosity = '0.25 0.25'
  []
  [rock_heat]
    type = PorousFlowMatrixInternalEnergy
    specific_heat_capacity = 800
    density = 2550
  [] 
  [rock_thermal_conductivity]
    type = PorousFlowThermalConductivityIdeal
    dry_thermal_conductivity = '2 0 0  0 2 0  0 0 2'
  []
[]


[DiracKernels]
  [heat_source]
    type = PorousFlowSquarePulsePointSource
    point = '1 1 0'
    mass_flux = 3e3
    variable = temperature
  []
[]
#[BCs]
#  [flux]
#    type = PorousFlowSink
#    boundary = 'left'
#    variable = temperature
#    use_mobility = false
#    use_relperm = false
#    mass_fraction_component = 1
#    fluid_phase = 0
#    flux_function = -3e3
#  []
#[]


[VectorPostprocessors]
  [vars]
    type = NodalValueSampler
    sort_by = id
    variable = 'pgas temperature'
    execute_on = 'timestep_end'
    outputs = spatial
  []
  [auxvars]
    type = ElementValueSampler
    sort_by = id
    variable = 'xAir swater'
    execute_on = 'timestep_end'
    outputs = spatial
  []
[]


[Postprocessors]
  [xAir]
    type = PointValue
    point =  '1 1 0'
    variable = xAir
  []
  [Pgas]
    type = PointValue
   point =  '1 1 0'
    variable = pgas
  []
  [T]
    type = PointValue
    point = '1 1 0'
    variable = temperature
  []
  [swater]
    type = PointValue
    point =  '1 1 0'
    variable = swater
  []
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]


[Executioner]
  type = Transient
  solve_type = NEWTON
  dt = 1e4
  num_steps = 1e6
  end_time = 3.56E8
  nl_abs_tol = 1e-20
  nl_max_its = 25
  l_max_its = 100
  dtmax = 5e6
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 1e3
  []
[]

[Outputs]
  exodus = true
  sync_times = '1e5 2.592e6 3.64e7 3e8'
  [time]
    type = CSV
  []
  [spatial]
    type = CSV
    sync_only = true
  []
[]
```

**Reply by batodon** [reply]
Hi @cpgr, any leads on this, pls? I hit a snag :(

**Reply by cpgr** [reply]
I seem to be hitting the same snag! I'll keep trying to see what is going on.

**Reply by batodon** [reply]
Thanks so much, @cpgr, for this feedback. I thought I was doing something wrong. I’ll now focus on why this convergence issue, and hopefully, I can get convergence and a true solution. Please, keep me in the loop if you’re successful.

**Reply by cpgr** [reply]
Just a note to say I'm still yet to figure this out!

**Reply by batodon** [reply]
Hello @cpgr, Thanks for the update! I was just about to update you myself :).
it seems the simulation is approaching the true solution despite not converging. I attached (here) the results after `1e6` seconds. One thing I noticed (which might be the error) is that `PorousFlow` fails to extrapolate beyond `110℃` and gives a convergence error beyond this value, which is a bit strange. While I'm still figuring out the issue, do you think using the `IdealGasAirFluidProperties` and `PorousFlowWaterAir` as TOUGH2’s eos4 in PorousFlow simulations is okay? Thank you. 

![EOS4_new](https://user-images.githubusercontent.com/113298585/223277465-17847987-3f79-46e7-ae78-637a19907e35.PNG)
